sklearn/decomposition/tests/test_dict_learning.py

```diff
@@ -366,3 +367,22 @@ def test_sparse_coder_estimator():
                        transform_alpha=0.001).transform(X)
     assert_true(not np.all(code == 0))
     assert_less(np.sqrt(np.sum((np.dot(code, V) - X) ** 2)), 0.1)
+
+
+def test_sparse_coder_parallel_mmap():
+    # Non-regression test for:
+    # https://github.com/scikit-learn/scikit-learn/issues/5956
+    # Test that SparseCoder does not error by passing reading only
+    # arrays to child processes
+
+    rng = np.random.RandomState(777)
+    n_components, n_features = 40, 64
+    init_dict = rng.rand(n_components, n_features)
+    # Ensure that `data` is >2M. Joblib memory maps arrays
+    # if they are larger than 1MB. The 4 accounts for float32
+    # data type
+    n_samples = int(2e6) // (4 * n_features)
+    data = np.random.rand(n_samples, n_features).astype(np.float32)
```

Comment:
Alternatively can just do `data.setflags(write=False)`. That way having a lot of data is not necessary to generate the same problem. Did something similar in PR ( https://github.com/scikit-learn/scikit-learn/pull/11342 ) in case it helps.

Reply:
Unfortunately, I believe it's not possible to use this as in this particular case it's an internal data structure `Xy` derived from the data that is being mutated.

We could write a unittest for the private function `_gram_omp` tough.